### ISSUE TITLE
Make `List` component usage more natural with shape of APIs response payload

### ISSUE THREAD

**Issue by DrewML** [magento-research/pwa-studio#41]
## This issue is for the following packages:

[ ] `venia-concept`
[ ] `pwa-buildpack`
[X] `peregrine`
[ ] `pwa-module`

## This issue is a:

[X] Bug
[X] Feature suggestion
[ ] Other (Please Specify)

### Environment

| Question                    |    Answer     |
| --------------------------- | :-----------: |
| Magento version             | n/a |
| Operating System + version  | n/a |
| node.js version (`node -v`) | n/a |
| npm version (`npm -v`)      | n/a |

## Description
The `List` component in Peregrine currently accepts an `items` prop, which is defined as:

```js
PropTypes.oneOfType([
   PropTypes.instanceOf(Map),
    PropTypes.arrayOf(PropTypes.array)
]).isRequired
```

The [call to `Array.from`](https://github.com/magento-research/pwa-studio/blob/434b895cec413258ea181590b007b508ecedf1cf/packages/peregrine/src/List/items.js#L61) masks an important error. If you pass in data straight from the API, which is in the shape:

```js
[{ id: 1 }, { id: 2 }]
```

you'll get an error about trying to destructure something that isn't an iterable. This is because `Array.from(exampleArrayAbove)` translates to a multi-dimensional array where each nested array has an Object as the value, but our code attempts to destructure `[key, item]` from the collection.

## Expected result:
I would expect, now that we know the shape of the data from the API, that we would want to make the expected shape in our components match 1:1.

## Possible solutions:
I'd like to switch to just accepting an array of objects for the sake of simplicity. We did not know the interface shape from GQL before creating this component, but we do now. We should optimize for the common case, and let people transform `Map` >> `Array` if they're doing some transformations from `Array` >> `Map` with the API response.

**Comment by jimbo** [comment]
Thanks for opening the issue. I'll walk through the design first, and then we can look at alternatives.

**Keys**

Like the `Map` constructor, the `List` component expects an iterable whose elements are key-value pairs (arrays with two elements). Like a `Map`, the collection of components returned by a `List` must be both indexed (order may be important) and keyed (each item **must** be accompanied by a unique key).

If you have a collection of data that is not keyed, such as a one-dimensional array of objects, that collection can't be transformed into a keyed list of components unless you do one of two things:

1. Transform the collection into a keyed one before passing it in
2. Pass a transform function that can be used to generate keys

At present, `List` is designed for the former.

**Data**

The typical shape of GraphQL data is the same as most API data: one-dimensional arrays of objects. Most APIs don't return keyed collections. Generally, there's a property on each object that serves as a primary key—typically `id`—but it does vary between data types. Also, some data types are flat arrays of literals, such as strings. Finally, `List` is used to map static data as well as API data.

What this means is that the transform function used to generate keys can't just be `_.pluck().`

**Array.from()**

All that `Array.from()` does is map an iterable to an array (of components) without generating an intermediate array. If you provide an iterable to `List` that matches the expected shape (compatible with `Map` constructor), it will be mapped into components correctly. If you don't, it will throw an error.

**Comment by DrewML** [comment]
Thanks! re: `Array.from`, I'm familiar with how it works 😸. The main point of this issue is calling out that we're not necessarily writing the world's most generic components - these components are designed for ease of use in PWA Studio. With that in mind, it would be ideal if, in most cases, you didn't have to make transformations to the data from the API. Less transformation code we have to ship down to the client.

It would be nice if a dev could look at the GraphQL schema and assume that the shape of something === what a related component expects.

**Comment by jimbo** [comment]
> With that in mind, it would be ideal if, in most cases, you didn't have to make transformations to the data from the API.

Right. If the API returned keyed collections, you wouldn't have to. Since it doesn't (and realistically never would), it has to be transformed by the consumer. As I pointed out before, the consumer has to either transform the data before passing it down, or write a transformer and give that to the component.

Are you proposing that we switch from the former to the latter?

**Comment by DrewML** [comment]
I guess what I'm getting at (and probably explaining poorly) is that it doesn't appear like this component ever needs to do a _lookup_ based on keys, it just needs a key for rendering. In that case, wouldn't a prop type expecting an array of object's with a unique `id` key be sufficient?

**Comment by jimbo** [comment]
The keys do need to be primary keys. Each item needs a unique key, and the list needs to know how to get the key for each object. It applies the key as a prop, and it may store the key in state for selection and other features.

Expecting an array of objects with an `id` prop would be sufficient for the most common case. For other cases, such as arrays of strings (e.g., tags) or arrays of objects with a different primary key (i.e., not `id`), again, either:

1. The collection would have to be transformed first
2. A function would have to be provided to determine the key

Originally, `List` accepted an array of `any`, and a function `getItemKey` that determined the key for each item. `getItemKey` defaulted to `({ id }) => id`, which optimized for the common case, as you're suggesting. We can return to that implementation if you'd like.

**Comment by jimbo** [comment]
Fixed by #62.
